pallets/reward-distribution/src/lib.rs

```diff
@@ -94,10 +159,212 @@ pub mod pallet {
+		if reward_currency_id == native_currency_id {
+			let available_native_funds = T::Balances::free_balance(&Self::fee_pool_account_id());
+
+			match reward.checked_sub(&available_native_funds) {
+				None => {
+					// Pay the whole reward from the fee pool
+					let amount: currency::Amount<T> = Amount::new(reward, reward_currency_id);
+					return amount.transfer(&Self::fee_pool_account_id(), &beneficiary)
+				},
+				Some(remaining) => {
+					//check if the to-be-minted amount is consistent
+					//with reward tracking
+					let liability = NativeLiability::<T>::get()
+						.ok_or(Error::<T>::NotEnoughRewardsRegistered)?;
+
+					if liability < remaining {
+						return Err(Error::<T>::NotEnoughRewardsRegistered.into())
+					}
+
+					// Use the available funds from the fee pool
+					let available_amount: currency::Amount<T> =
+						Amount::new(available_native_funds, reward_currency_id);
+					available_amount.transfer(&Self::fee_pool_account_id(), &beneficiary)?;
+					// Mint the rest
+					T::Balances::deposit_creating(&beneficiary, remaining);
+
+					NativeLiability::<T>::set(Some(liability - remaining));
+				},
+			}
+			Ok(())
+		} else {
+			//we need no checking of available funds, since the fee pool will ALWAYS have enough
+			// collected fees of the wrapped currencies
+			let amount: currency::Amount<T> = Amount::new(reward, reward_currency_id);
+			amount.transfer(&Self::fee_pool_account_id(), &beneficiary)
+		}
+	}
+
+	pub fn execute_on_init(_height: T::BlockNumber) {
+		//get reward per block
+		let reward_per_block = match RewardPerBlock::<T>::get() {
+			Some(value) => value,
+			None => {
+				log::warn!("Reward per block is None");
+				return
+			},
+		};
+
+		if let Err(_) = ext::security::ensure_parachain_status_running::<T>() {
+			return
+		}
+
+		let rewards_adapted_at = match RewardsAdaptedAt::<T>::get() {
+			Some(value) => value,
+			None => {
+				log::warn!("RewardsAdaptedAt is None");
+				return
+			},
+		};
+
+		let mut reward_this_block = reward_per_block;
+
+		//update the reward per block if decay interval passed
+		if let Ok(expired) = ext::security::parachain_block_expired::<T>(
+			rewards_adapted_at,
+			T::DecayInterval::get() - T::BlockNumber::one(),
+		) {
+			if expired {
+				let decay_rate = T::DecayRate::get();
+				reward_this_block = (Perquintill::one() - decay_rate).mul_floor(reward_per_block);
+				RewardPerBlock::<T>::set(Some(reward_this_block));
+				let active_block = ext::security::get_active_block::<T>();
+				RewardsAdaptedAt::<T>::set(Some(active_block));
+			}
+		} else {
+			log::warn!("Failed to check if the parachain block expired");
+		}
+		//keep track of total native currencies to be minted
+		NativeLiability::<T>::mutate(|current_liability| match current_liability {
+			Some(current_liability) => *current_liability += reward_this_block,
+			None => *current_liability = Some(reward_this_block),
+		});
+		//TODO how to handle error if on init cannot fail?
```

Comment:
Maybe we should 
```
if let Err(err) = Self::distribute_rewards(...)
```
and try to log the error message? Not much we can do otherwise.

Reply:
With a trace. Yes, I will do that!